docs/man/coherence.1
.\" Man page generated from reStructeredText.
.
.TH COHERENCE 1 "Thu, Mar 19 2009" "Version 0.6.7" ""
.SH NAME
coherence \- Python UPnP/DLNA Media Server and Framework
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" -*- mode: rst -*-
.
.\" disable justification (adjust text to left margin only)
.ad l
.SH SYNOPSIS
.sp
\fBcoherence\fP <options> [\-\-plugin=<BACKEND> [ , <PARAM_NAME> : <PARAM_VALUE> ] ...]
.SH DESCRIPTION
.sp
Coherence is a Python UPnP framework which enabling your application to
participate in digital living networks, at the moment primarily the
UPnP universe.
.sp
Its goal is to relieve your application from all the membership and UPnP
related tasks as much as possible.
.sp
The core of Coherence provides a (hopefully complete) implementation
of:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
.
a SSDP server,
.IP \(bu 2
.
a MSEARCH client,
.IP \(bu 2
.
server and client for HTTP/SOAP requests, and
.IP \(bu 2
.
server and client for Event Subscription and Notification (GENA).
.UNINDENT
.UNINDENT
.UNINDENT
.SH OPTIONS
.INDENT 0.0
.TP
.B \-v,  \-\-version
.
Show program\(aqs version number and exit
.TP
.B \-\-help
.
Show help message and exit
.TP
.B \-d,  \-\-daemon
.
Daemonize
.TP
.BI \-c,  \-\-configfile\fB= PATH
.
Path to config file
.TP
.B \-\-noconfig
.
ignore any config file found
.TP
.BI \-o,  \-\-option\fB= OPTION
.
activate option
.TP
.BI \-l,  \-\-logfile\fB= PATH
.
Path to log file.
.UNINDENT
.SH EXAMPLES
.INDENT 0.0
.TP
.B coherence \-\-plugin=backend:FSStore,name:MyCoherence
.
Start coherence activating the \fIFSStore\fP backend.
.TP
.B coherence \-\-plugin=backend:MediaStore,medialocation:$HOME/Music/,mediadb:/tmp/media.db
.
Start coherence activating the \fIMediaStore\fP backend with media
located in \fI$HOME/Music\fP and the media metadata store in
\fI/tmp/media.db\fP.
.UNINDENT
.SH AVAILABLE STORES
.sp
BetterLight, AmpacheStore, FlickrStore, MiroStore, ElisaPlayer,
ElisaMediaStore, Gallery2Store, DVBDStore, FSStore, BuzztardPlayer,
BuzztardStore, GStreamerPlayer, SimpleLight, ITVStore, SWR3Store,
TrackerStore, LolcatsStore, BBCStore, MediaStore, AppleTrailerStore,
LastFMStore, AxisCamStore, YouTubeStore, TEDStore, IRadioStore
.SH FILES
.INDENT 0.0
.TP
.B $HOME/.coherence
.
default config file
.UNINDENT
.SH ENVIRONMENT VARIABLES
.INDENT 0.0
.TP
.B COHERENCE_DEBUG=<STORE>
.
Supplies debug information pertaining to the named store.
.UNINDENT
.SH SEE ALSO
.sp
Project Homepage \fI\%http://coherence.beebits.net/\fP
.SH AUTHOR
Coherence was written by Frank Scholz
<coherence@beebits.net>. This man page was created by
Charlie Smotherman <cjsmo@cableone.net> for
Frank Scholz and the Debian Project.

Licence: GNU Public Licence v3 (GPLv3)
.\" Generated by docutils manpage writer.
.\"
.


NEWS
0.6.6.2 - Red-Nosed Reindeer Revolutions - 20100102

 * fixing the setup configuration to compensate for the gone
   'auto-include-all-files-under-version-control' setuptools feature 
 * generic 'genre' handling in DIDLLite - thx Caleb
 * fix for broken .face-icon handling in the Rhythmbox plugin

0.6.6.1 - Red-Nosed Reindeer Reloaded - 20091222

 * fixing the broken tarball of the 0.6.6 release and
   adding a missing import that made some backends unfeasible

0.6.6 - Red-Nosed Reindeer - 20091220

 * new MediaServer backends that allow access to
   * Banshee - exports audio and video files from Banshees media db
     (http://banshee-project.org/)
   * FeedStore - a MediaServer serving generic RSS feeds
   * Playlist - exposes the list of video/audio streams from a m3u playlist
     (e.g. web TV listings published by french ISPs such as Free, SFR...)
   * YAMJ - serves the movie/TV series data files and metadata
     from a given YAMJ (Yet Another Movie Jukebox) library
     (http://code.google.com/p/moviejukebox/)
 * updates on Mirabeau - our "UPnP over XMPP" bridge
 * simplifications in the D-Bus API
 * a first implementation of an JSON/REST API
 * advancements of the GStreamer MediaRenderer, supporting now GStreamers playbin2
 * upgrade of the DVB-Daemon MediaServer
 * refinements in the transcoding section, having now the choice to use GStreamer pipelines
   or external processes like mencoder
 * more 'compatibility' improvements for different devices (e.g. Samsung TVs or Apache Felix)
 * and - as every time - the usual bugfixes and enhancements

Kudos go to Benjamin (lightyear) Kampmann,
            Dominik (schrei5) Ruf,
            Frank (dev) Scholz,
            Friedrich (frinring) Kossebau,
            Jean-Michel (jmsizun) Sizun,
            Philippe (philn) Normand,
            Sebastian (sebp) Pölsterl,
            Zaheer (zaheerm) Merali

0.6.4 - Pont Mirabeau - 20090512

 * new MediaServer backends that allow access to
   * Picasa Web Albums (http://picasa.google.com)
   * a TestServer to easily serve and test interaction with
     * one or more items and adjust 'upnp_class', 'mimetype' and 'DLNA-flags',
     * items that are a GStreamer pipeline or an external program
 * a new - used in parallel - D-Bus API with an 'org.DLNA' interface
   with the goal to create a common API for all UPnP/DNLA frameworks
 * support for the dlna-playcontainer URI
   (http://netzflocken.de/2009/4/23/media-collection-playing-the-dlna-way)
 * enchancements to the GStreamer MediaRenderer, supporting now
   dlna-playcontainer and SetNextAVTransportURI, and jumping to previous
   and next tracks
 * support for video items served by Ampache (http://ampache.org)
 * base classes for a ScheduledRecording service
 * some 'compatibility' adjustments for different devices
 * and - as every time - the usual bugfixes and enhancements

   Kudos go to jmsizun, Stecchino, cjsmo, chewi, and lightyear.

0.6.2 - Rosenmontag - 20090223

 * new MediaServer backends that allow access to
   * YouTube videos (http://youtube.com)
   * the MiroGuide for online videos (https://www.miroguide.com)
   * the videos provided by Shoutcast TV (http://www.shoutcast.com)
   * the SWR3 podcasts, a German radio station (http://swr3.de)
 * adjustments to the Ampache backend to work with newer Ampache versions (http://ampache.org)
 * a lot of 'compatibility' enhancements for different devices
 * a 'port' to the OpenEmbedded platform (http://www.openembedded.org/),
   bringing Coherence to the BeagleBoard (http://beagleboard.org/)
 * and - as every time - the usual bugfixes and enhancements

   Kudos go especially to jmsizun for his work on the new backends!

0.6.0 - The late Pumpkin - 20081231

 * new MediaServer backends that allow access to
   * movie trailers from the Apple HD trailers site (http://www.apple.com/trailers)
   * images hosted at a Gallery site - an open source web based photo album organizer (http://gallery.menalto.com)
   * Lolcats images from http://icanhascheezburger.com
   * podcasts from the BBC (http://open.bbc.co.uk/labs/)
   * videos from TED (http://www.ted.com)
 * an extended Flickr MediaServer backend
   * enables user-authenticated access to your Flickr account
   * access to your images and the one of your frieds via an UPnP device
   * upload an image directly from an UPnP device to your Flickr account
 * transcoding of audio files based on GStreamer for DLNA devices like the PS3, and even XBox 360
 * several plugins for the Nautlilus filemanager (http://www.gnome.org/projects/nautilus)
   * sharing folders from within Nautilus
   * upload files from Nautilus to UPnP A/V MediaServers
   * play files from Nautilus on an UPnP A/V MediaRenderer
 * an experimental plugin for EOG - the Gnome Image Viewer (http://projects.gnome.org/eog/)
 * greatly improved XBox 360 support, including audio transcoding
 * and the usual bugfixes and enhancements
   Kudos go especially to jmsizun, lightyear, superdump for their work
   on the backends and their patient debugging sessions!

0.5.8 - Trix and Flix - 20080630

 * a MediaServer backend for DVB-Daemon (http://www.k-d-w.org/node/42)
   * exporting atm the stored recordings
   * allowing to delete recordings from within a UPnP client, when enabled on the backend
   * will export EPG data and allow scheduling via UPnP in the future
 * client device and service implementations for BinaryLight and DimmableLight devices
 * rework of the D-Bus support
   * should now be usable from other languages (C,Perl,..) too
   * support for activating/deactivation a backend via D-Bus, allowing for instance to start a MediaServer backend via D-Bus
 * a plugin for Totem (http://www.gnome.org/projects/totem/)
   * enabling Totem to detect and browse UPnP A/V MediaServers
   * using only D-Bus to communicate with a Coherence instance
 * a basic reusable PyGTK based UPnP A/V ControlPoint widget, used in the Totem plugin
 * rework (again) of the XBox 360 support - getting closer

coherence.py
#! /usr/bin/env python
#
# Licensed under the MIT license
# http://opensource.org/licenses/mit-license.php

# Copyright 2006, Frank Scholz <coherence@beebits.net>

import string
import socket
import os, sys

from zope.interface import implements, Interface
from twisted.python import log, filepath, util
from twisted.python.components import registerAdapter
from twisted.internet import task, address
from twisted.internet import reactor
from nevow import athena, inevow, loaders, tags, static
from twisted.web import resource

import louie

from coherence.upnp.core.ssdp import SSDPServer
from coherence.upnp.core.msearch import MSearch
from coherence.upnp.core.device import Device, RootDevice
from coherence.upnp.core.utils import parse_xml

from coherence.extern.logger import Logger
log = Logger('Coherence')

class IWeb(Interface):

    def goingLive(self):
        pass

class Web(object):

    def __init__(self, coherence):
        super(Web, self).__init__()
        self.coherence = coherence

class WebUI(athena.LivePage):
    """
    """

    addSlash = True
    docFactory = loaders.xmlstr("""\
<html xmlns:nevow="http://nevow.com/ns/nevow/0.1">
<head>
<nevow:invisible nevow:render="liveglue" />
<link rel="stylesheet" type="text/css" href="main.css" />
</head>
<body>
Coherence - a Python UPnP A/V framework
<p>
<div nevow:render="listchilds"></div>
</p>
</body>
</html>
""")

    def __init__(self, *a, **kw):
        super(WebUI, self).__init__( *a, **kw)
        self.coherence = self.rootObject.coherence

    def childFactory(self, ctx, name):
        log.info('WebUI childFactory: %s' % name)
        try:
            return self.rootObject.coherence.children[name]
        except:
            ch = super(WebUI, self).childFactory(ctx, name)
            if ch is None:
                p = util.sibpath(__file__, name)
                if os.path.exists(p):
                    ch = static.File(p)
            return ch
        
    def render_listchilds(self, ctx, data):
        cl = []
        log.info('children: %s' % self.coherence.children)
        for c in self.coherence.children:
            device = self.coherence.get_device_with_id(c)
            if device != None:
                _,_,_,device_type,version = device.get_device_type().split(':')
                cl.append( tags.li[tags.a(href='/'+c)[device_type,
                                                      ':',
                                                      version,
                                                      ' ',
                                                      device.get_friendly_name()]])
            else:
                cl.append( tags.li[c])
        return ctx.tag[tags.ul[cl]]
        
class WebServer:

    def __init__(self, port, coherence):
        from nevow import appserver

        def ResourceFactory( original):
            return WebUI( IWeb, original)

        registerAdapter(ResourceFactory, Web, inevow.IResource)

        self.web_root_resource = Web(coherence)
        #self.web_root_resource = inevow.IResource( web)
        #print self.web_root_resource
        self.site = appserver.NevowSite( self.web_root_resource)
        reactor.listenTCP( port, self.site)
        
        log.msg( "WebServer on port %d ready" % port)


class Coherence:

    def __init__(self):
        self.enable_log = False
        self.devices = []
        
        self.children = {}
        self._callbacks = {}
        
        logmode = 'info'        # FIXME: get this from the config file
                                #               and have one for every module
        self.enable_log = True  # FIXME: set this to True if a logmode != 'none'
        
        log.disable(name='Coherence')
        log.disable(name='SSDP')
        log.disable(name='MSEARCH')
        log.disable(name='Service')
        log.disable(name='SOAP')
        log.disable(name='MediaServer')
        log.disable(name='MediaRenderer')
        log.disable(name='Event')
        
        log.enable(name='Coherence')
        #log.enable(name='MediaServer')
        log.set_level(name='MediaServer')
        #log.enable(name='MediaRenderer')
        log.set_level(name='MediaRenderer')
        #log.enable(name='SOAP')
        log.set_level(name='SOAP')
        #log.enable(name='Service')
        log.set_level(name='Service')
        #log.enable(name='MSEARCH')
        log.set_level(name='MSEARCH')
        #log.enable(name='Event')
        log.set_level(name='Event')
        
        plugin = louie.TwistedDispatchPlugin()
        louie.install_plugin(plugin)

        log.msg("Coherence UPnP framework starting...")
        self.ssdp_server = SSDPServer()
        louie.connect( self.add_device, 'Coherence.UPnP.SSDP.new_device', louie.Any)
        louie.connect( self.remove_device, 'Coherence.UPnP.SSDP.remove_device', louie.Any)
        louie.connect( self.receiver, 'Coherence.UPnP.Device.detection_completed', louie.Any)
        #louie.connect( self.receiver, 'Coherence.UPnP.Service.detection_completed', louie.Any)

        self.ssdp_server.subscribe("new_device", self.add_device)
        self.ssdp_server.subscribe("removed_device", self.remove_device)

        self.msearch = MSearch(self.ssdp_server)

        reactor.addSystemEventTrigger( 'before', 'shutdown', self.shutdown)

        self.web_server_port = 30020
        self.hostname = socket.gethostbyname(socket.gethostname())
        #FIXME: this doesn't work on systems with more than one network interface
        log.msg('running on host: %s' % self.hostname)
        self.urlbase = 'http://%s:%d/' % (self.hostname, self.web_server_port)

        self.web_server = WebServer( self.web_server_port, self)
                                
        self.renew_service_subscription_loop = task.LoopingCall(self.check_devices)
        self.renew_service_subscription_loop.start(20.0, now=False)


    
        # are we supposed to start a ControlPoint?
        try:
            from coherence.upnp.devices.control_point import ControlPoint
            #ControlPoint( self)
        except ImportError:
            log.msg("Can't enable ControlPoint functions, sub-system not available.")

        
        # are we supposed to start a MediaServer?
        try:
            from coherence.upnp.devices.media_server import MediaServer
            MediaServer( self,version=2)
        except ImportError:
            log.msg("Can't enable MediaServer functions, sub-system not available.")


        # are we supposed to start a MediaRenderer?
        try:
            from coherence.upnp.devices.media_renderer import MediaRenderer
            MediaRenderer( self)
        except ImportError:
            log.msg("Can't enable MediaRenderer functions, sub-system not available.")
        
    def receiver( self, signal, *args, **kwargs):
        #print "Coherence receiver called with", signal
        #print kwargs
        pass

    def shutdown( self):
        """ send service unsubscribe messages """
        try:
            self.renew_service_subscription_loop.stop()
        except:
            pass
        for root_device in self.get_devices():
            root_device.unsubscribe_service_subscriptions()
            for device in root_device.get_devices():
                device.unsubscribe_service_subscriptions()
        self.ssdp_server.shutdown()
        log.msg('Coherence UPnP framework shutdown')
        
    def check_devices(self):
        """ iterate over devices and their embedded ones and renew subscriptions """
        for root_device in self.get_devices():
            root_device.renew_service_subscriptions()
            for device in root_device.get_devices():
                device.renew_service_subscriptions()

    def subscribe(self, name, callback):
        self._callbacks.setdefault(name,[]).append(callback)

    def unsubscribe(self, name, callback):
        callbacks = self._callbacks.get(name,[])
        if callback in callbacks:
            callbacks.remove(callback)
        self._callbacks[name] = callbacks

    def callback(self, name, *args):
        for callback in self._callbacks.get(name,[]):
            callback(*args)


    def get_device_with_usn(self, usn):
        found = None
        for device in self.devices:
            if device.get_usn() == usn:
                found = device
                break
        return found

    def get_device_with_id(self, device_id):
        found = None
        for device in self.devices:
            id = device.get_id()
            if device_id[:5] != 'uuid:':
                id = id[5:]
            if id == device_id:
                found = device
                break
        return found

    def get_devices(self):
        return self.devices

    def add_device(self, device_type, infos):
        if infos['ST'] == 'upnp:rootdevice':
            root = RootDevice(infos)
            self.devices.append(root)
        else:
            root_id = infos['USN'][:-len(infos['ST'])-2]
            root = self.get_device_with_id(root_id)
            device = Device(infos, root)
        # fire this only after the device detection is fully completed
        # and we are on the device level already, so we can work with them instead with the SSDP announce
        #if infos['ST'] == 'upnp:rootdevice':
        #    self.callback("new_device", infos['ST'], infos)


    def remove_device(self, device_type, infos):
        device = self.get_device_with_usn(infos['USN'])
        if device:
            self.devices.remove(device)
            if infos['ST'] == 'upnp:rootdevice':
                self.callback("removed_device", infos['ST'], infos['USN'])

            
    def add_web_resource(self, name, sub):
        #self.web_server.web_root_resource.putChild(name, sub)
        self.children[name] = sub
        #print self.web_server.web_root_resource.children

    def remove_web_resource(self, name):
        # XXX implement me
        pass
        
def main():

    # get settings or options
    Coherence()
    reactor.run()
    
if __name__ == '__main__':
    main()
